src/connectors/tests/clickhouse.rs

```diff
@@ -0,0 +1,167 @@
+use signal_hook_async_std::Signals;
+use testcontainers::{clients, core::Port, images::generic::GenericImage, Docker, RunArgs};
+use tremor_common::ports::IN;
+use tremor_pipeline::{Event, EventId};
+use tremor_script::literal;
+
+use super::ConnectorHarness;
+
+const CONTAINER_NAME: &str = "clickhouse/clickhouse-server";
+const CONTAINER_VERSION: &str = "22.3.3.44";
+const SERVER_PORT: u16 = 9000;
+
+const DB_HOST: &str = "127.0.0.1";
+
+#[async_std::test]
+async fn simple_insertion() -> Result<()> {
+    let _ = env_logger::try_init();
+
+    let docker = clients::Cli::default();
+    let image = GenericImage::new(format!("{CONTAINER_NAME}:{CONTAINER_VERSION}"));
+
+    // We want to access the container from the host, so we need to make the
+    // corresponding port available.
+    let port_to_expose = Port {
+        local: SERVER_PORT,
+        internal: SERVER_PORT,
+    };
+    let container =
+        docker.run_with_args(image, RunArgs::default().with_mapped_port(port_to_expose));
+
+    let container_id = container.id().to_string();
+    let mut signals = Signals::new(&[SIGTERM, SIGINT, SIGQUIT])?;
+    let signal_handle = signals.handle();
+    let signal_handler_task = async_std::task::spawn(async move {
+        let signal_docker = clients::Cli::default();
+        while let Some(_signal) = signals.next().await {
+            signal_docker.stop(container_id.as_str());
+            signal_docker.rm(container_id.as_str());
+        }
+    });
+    let port = container.get_host_port(9000).unwrap();
+
+    wait_for_ok(port).await.unwrap();
+
+    create_table(port, "people").await.unwrap();
+
+    let connector_config = literal!({
+        "config": {
+            "host": DB_HOST,
+            "port": port,
+            "database": "",
+        },
+    });
+
+    let harness = ConnectorHarness::new("clickhouse", &connector_config).await?;
+
+    let in_pipe = harness.get_pipe(IN).expect("No pipe connected to port IN");
+
+    harness.start().await?;
+    harness.wait_for_connected().await?;
+
+    let batched_data = literal!([
+        {
+            "data": {
+                "value": {
+                    "age": 42u8,
+                }
+            },
+        },
+        {
+            "data": {
+                "value": {
+                    "age": 101u8,
+                }
+            },
+        },
+    ]);
+
+    let batched_meta = literal!({});
+
+    let event = Event {
+        id: EventId::new(0, 0, 1, 1),
+        is_batch: true,
+        transactional: true,
+        data: (batched_data, batched_meta).into(),
+        ..Event::default()
+    };
+
+    harness.send_to_sink(event, IN).await?;
+    // TODO: check for some kind of ack or something i guess?
+
+    info!("SUCCESSFULLY INSERTED DATA");
+
+    // Now that we have sent some events to the sink, let's check that
+    // everything was properly inserted in the database.
+
+    harness.stop().await?;
+
+    // Please don't judge me too much - this is a yolo test.
```

Comment:
Haha! Don't worry :D

Reply:
https://github.com/tremor-rs/tremor-runtime/blob/main/src/connectors/tests/ws.rs#L547-L551

Perhaps adding a TestClickhouse client to drive `expect()` style tests as in the ws connector test would be helpful here?

Reply:
You could simply use a clickhouse client that checks in a loop until a certain timeout if the table already has the expected number of rows. If the table doesn't have the expected number within a certain timeout (say 10 seconds), then throw an error.